### DISCUSSION TITLE
Payment Integration Issues with PocketBase JS Hooks

### DISCUSSION THREAD

**Discussion by rustamous** [pocketbase/pocketbase#5907] [Q&A] [answered]
Hi,

I'm implementing YooKassa payment widget(https://yookassa.ru/developers/payment-acceptance/integration-scenarios/widget/basics) in a React + PocketBase application. I'm encountering persistent issues with JavaScript hooks and global configuration in PocketBase.

Project Structure:
React frontend with YooKassa embedded widget
PocketBase backend (pocketbase 0.22.21) handling payment creation
Payment flow implementation in pb_hooks/main.pb.js

Current Issues:
Inconsistent error messages when trying to define global configuration:

- ReferenceError: YOOKASSA_CONFIG is not defined
- TypeError: Object has no member 'set'
- TypeError: Cannot read property 'shop_id' of undefined
- TypeError: Object has no member 'crypto'
- ReferenceError: TextEncoder is not defined

**Code Implementation:**
/// <reference path="../pb_data/types.d.ts" />

// Define configuration as a global variable
const YOOKASSA_CONFIG = {
        shop_id: '488***',
        secret_key: 'test_***'
    };

Questions:
1. What's the recommended approach for storing sensitive configuration (YooKassa credentials) in PocketBase hooks?
2. Is there a built-in way to handle Base64 encoding in PocketBase JS VM? Getting ReferenceError: btoa is not defined
3. How should we properly initialize and access global variables in pb_hooks?

The errors seem to rotate between different JavaScript runtime limitations in the PocketBase JS VM. Looking for guidance on the proper way to implement this payment integration within PocketBase's constraints.

Best regards

**Comment by ganigeorgiev** [comment, marked as answer]
Without a more complete code sample, it is not clear where the config variable is defined but for the handler scope limitations please check https://pocketbase.io/docs/js-overview/#handlers-scope.

One possible option in your case could be to save your keys as plain `.json` file and then simply import it with ``const config = require(`${__hooks}/config.json`)``.

Regarding the btoa error - atob/btoa functions are not available at the moment. I'll consider enabling the `Buffer` goja_nodejs module with the upcoming v0.23.0, but for now you'll have to reimplement them or load a polyfill on your own.

**Reply by rustamous** [reply]
Thank you, 
It really helped!
`const config = require(`${__hooks}/config.json`).`
` but for now you'll have to reimplement them or load a polyfill on your own.`
